Try to fix IRQ vector table relocation

carlocaione opened the pull request:
The generation of the software ISR table and the IRQ vector table (respectively generated by `CONFIG_GEN_SW_ISR_TABLE` and `CONFIG_GEN_IRQ_VECTOR_TABLE`) should (in theory) go through three stages:
    
1. A placeholder table is generated in `arch/common/isr_tables.c` and placed in an orphaned `.gnu.linkonce.{irq_vector_table, sw_isr_table}` section
    
2. The real table is generated by `arch/common/gen_isr_tables.py` (creating the `build/zephyr/isr_tables.c` file)
    
3. The real table is un-orphaned by moving it in a proper section with a proper alignment
  
While all the steps are done automatically for the software ISR table, for the IRQ vector table each architecture must take care of modifying its own linker script to place somewhere the generated IRQ vector table (basically step 3 is missing).
    
This is currently only done for 2 architectures: ARM and ARC. But when another architecture tries to use the IRQ vector table, the linker complains about that. For example:

 ```
      Linking C executable zephyr/zephyr.elf
      riscv64-zephyr-elf/bin/ld.bfd: warning: orphan section
        `.gnu.linkonce.irq_vector_table' from
        `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in
        section `.gnu.linkonce.irq_vector_table'
```

In this patchset we introduce a new `CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN` to support the architectures requiring a special alignment for the IRQ vector table and we also introduce a way to automatically place the IRQ vector table in place in the same way it is done for the ISR software table.

We also convert the ARM and ARC architectures to use this new mechanism and remove the support on ARM64 where the IRQ vector table cannot be used.

carlocaione committed:
irq: Fix IRQ vector table relocation

The generation of the software ISR table and the IRQ vector table
(respectively generated by CONFIG_GEN_SW_ISR_TABLE and
CONFIG_GEN_IRQ_VECTOR_TABLE) should (in theory) go through three stages:

1. A placeholder table is generated in arch/common/isr_tables.c and
   placed in an orphaned .gnu.linkonce.{irq_vector_table, sw_isr_table}
   section

2. The real table is generated by arch/common/gen_isr_tables.py (creating
   the build/zephyr/isr_tables.c file)

3. The real table is un-orphaned by moving it in a proper section with a
   proper alignment

While all the steps are done automatically for the software ISR table,
for the IRQ vector table each architectures must take care of modiying
its own linker script to place somewhere the generated IRQ vector table
(basically step 3 is missing).

This is currently only done for 2 architectures: Cortex-M (ARMv7) and
ARC. But when another architecture tries to use the IRQ vector table,
the linker complains about that. For example:

  Linking C executable zephyr/zephyr.elf
  riscv64-zephyr-elf/bin/ld.bfd: warning: orphan section
    `.gnu.linkonce.irq_vector_table' from
    `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in
    section `.gnu.linkonce.irq_vector_table'

In this patch we introduce a new CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN to
support the architectures requiring a special alignment for the IRQ
vector table and we also introduce a way to automatically place the IRQ
vector table in place in the same way it is done for the ISR software
table.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arm64: linker: Remove IRQ vector table symbols

The IRQ vector table cannot be used on this architecture.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arc: vector_table: Automatically place the IRQ vector table

Instead of using a custom linker script, rely on the automatic placement
of the IRQ vector table.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arm: vector_table: Automatically place the IRQ vector table

Instead of using a custom linker script, rely on the automatic placement
of the IRQ vector table.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
test: arm_irq_vector_table: Fix when !CONFIG_GEN_IRQ_VECTOR_TABLE

This test is based on the wrong assumption that the IRQ vector table
symbols are always placed by the linker script also when
CONFIG_GEN_IRQ_VECTOR_TABLE is not actually set.

This is of course broken with the reworked mechanism of the IRQ vector
table placement.

Fix the test by reintroducing the old behaviour by forcefully
reinstating the symbols in the linker script.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

andyross reviewed (dismissed):
Seems very reasonable.

microbuilder commented:
@carlocaione Thanks for the useful cleanup. Looks good for Arm. 👍

evgeniy-paltsev commented on include/zephyr/linker/irq-vector-table-section.ld:
```diff
@@ -0,0 +1,4 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+. = ALIGN(CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN);
```
Please correct me, but are we sure that this change isn't introduce regression? Previously we align _IRQ_VECTOR_TABLE_SECTION_SYMS with default section alignment (at least), but now we have `CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN` with default value '0' - so now we have
`. = ALIGN(0);`

Not fully sure what linker will do with `ALIGN(0)`

andyross replied:
No one is fully sure what a linker is going to do with anything.  But I'd be very surprised if this blew up.

ruuddw replied:
It doesn't sound too complex to find out what happens. I assume for gcc we have CI coverage. Evgeniy, your concern is about the llvm linker?

carlocaione replied:
@evgeniy-paltsev
> Please correct me, but are we sure that this change isn't introduce regression? 

"the only certainty in life is death and taxes" for all the rest there is CI, and CI is all green. If a new arch needs some special alignment (like RISCV), then the newly introduced `CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN` can be used.

For all the other architectures FWICT the alignment of the generated IRQ vector table is not important (please note that this vector table is not the same as https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/aarch32/cortex_m/vector_table.S, just to avoid confusion)

> Previously we align _IRQ_VECTOR_TABLE_SECTION_SYMS with default section alignment (at least)

For ARM and ARC the vector table is placed after `.exc_vector_table.*` symbols, so it is really not aligned to anything defined (and this is one reason indeed to introduce `CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN`)

Also please consider that this PR is not touching the address where the `_irq_vector_table` is placed: the `vector_table.ld` file is placed with key `0x0vectors`, the same sort key is used to place the `irq-vector-table-section.ld`, so nothing is changing (as long as there are no `.vectors` symbols, and there are not).

> Not fully sure what linker will do with ALIGN(0)

I'm not inventing anything new here, this is already used in Zephyr, see https://github.com/zephyrproject-rtos/zephyr/blob/f79eaba7e6d2c698fbaba3f7bc7348f9a78a611d/arch/Kconfig#L421-L423
https://github.com/zephyrproject-rtos/zephyr/blob/f79eaba7e6d2c698fbaba3f7bc7348f9a78a611d/include/zephyr/linker/common-ram.ld#L25-L26
so I would assume this is safe.

evgeniy-paltsev replied:
Ok, I've run verification with MWDT toolchain, though maybe it's better to set CONFIG_ARCH_SW_ISR_TABLE_ALIGN to 1 by default instead of 0?

So we get 
```
. = ALIGN(1); 
```
instead of
```
. = ALIGN(0); 
```
by default. At least it's more meaningful and I expect it should be more acceptable for all linkers.

@carlocaione  What do you think?

carlocaione replied:
I think that at this point we should set a more proper value for both `ARCH_IRQ_VECTOR_TABLE_ALIGN` and `ARCH_SW_ISR_TABLE_ALIGN` that it should probably be `4` to have a 32-bit alignment.

carlocaione committed:
arch: Use a more sane ALIGN value

By default ARCH_IRQ_VECTOR_TABLE_ALIGN and ARCH_SW_ISR_TABLE_ALIGN are
set to 0. Use a more proper value.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@microbuilder @andyross can you ACK this again if you want?
